zerver/actions/message_send.py

```diff
@@ -972,6 +975,36 @@ def do_send_messages(
                     )
                     send_request.automatic_new_visibility_policy = new_visibility_policy
 
+            # Set the visibility_policy of the users mentioned in the message
+            # to "FOLLOWED" if "automatically_follow_topics_where_mentioned" is "True".
+            user_mentions = (
+                send_request.rendering_result.mentions_user_ids & send_request.active_user_ids
+            )
+            user_profiles_query_set = UserProfile.objects.filter(id__in=user_mentions)
+            user_profiles = set()
+            user_topics_already_followed = set()
+            for user_profile in user_profiles_query_set:
+                if user_profile.automatically_follow_topics_where_mentioned:
+                    user_profiles.add(user_profile)
+
+            user_topics = UserTopic.objects.filter(
+                user_profile__in=user_profiles,
+                stream_id=send_request.stream.id,
+                topic_name__iexact=send_request.message.topic_name(),
+            )
+            for user_topic in user_topics:
+                if user_topic.visibility_policy == UserTopic.VisibilityPolicy.FOLLOWED:
+                    user_topics_already_followed.add(user_topic.user_profile)
+
```

Comment:
Better query would be
```py
UserProfile.objects.filter(id__in=user_mentions, automatically_follow_topics_where_mentioned=True)
 ```
 and
 
 ```py
             user_topics = UserTopic.objects.filter(
                user_profile__in=user_profiles,
                stream_id=send_request.stream.id,
                topic_name__iexact=send_request.message.topic_name(),
            ).exclude(visibility_policy = UserTopic.VisibilityPolicy.FOLLOWED)
  ```
 We'll be able to skip these loops, right?

Reply:
This looks good but there will be one loop to get user_profiles from user_topics.

Reply:
Can you explain? -- or it'd be better to push your code with the changes for better understanding.

Reply:
I have pushed changes in my recent commit. You can take look into that. I have not use ```exclude(visibility_policy = UserTopic.VisibilityPolicy.FOLLOWED)``` because it will not contain the users where UserTopic.DoesNotExist occurs. Instead I have removed the users who have already followed the topic from the mentioned users.